src/network/networkctl.c

```diff
@@ -242,6 +249,64 @@ static const LinkInfo* link_info_array_free(LinkInfo *array) {
 }
 DEFINE_TRIVIAL_CLEANUP_FUNC(LinkInfo*, link_info_array_free);
 
+DEFINE_TRIVIAL_CLEANUP_FUNC(CURL*, curl_easy_cleanup);
+DEFINE_TRIVIAL_CLEANUP_FUNC(struct curl_slist*, curl_slist_free_all);
+
+static int acquire_azure_cloud_metadata(NetworkCloudManager **ret) {
+        _cleanup_(curl_slist_free_allp) struct curl_slist *request_header = NULL;
+        _cleanup_(network_cloud_manager_freep) NetworkCloudManager *m = NULL;
+        _cleanup_(json_variant_unrefp) JsonVariant *v = NULL;
+        _cleanup_(curl_easy_cleanupp) CURL *curl = NULL;
+        _cleanup_free_ char *url = NULL;
+        int r;
+
+        if (detect_virtualization() != VIRTUALIZATION_MICROSOFT)
+                return -ENOTSUP;
+
+        r = network_cloud_manager_new(CLOUD_PROVIDER_AZURE, &m);
+        if (r < 0)
+                return r;
+
+        curl = curl_easy_init();
+        if (!curl)
+                return -ENOSR;
+
+        url = strjoin("http://", AZURE_IMDS_REST_ENDPOINT, AZURE_NETWORK_METADATA_URL_BASE, AZURE_API_VERSION);
+        if (!url)
+                return log_oom();
+
+        if (curl_easy_setopt(curl, CURLOPT_URL, url) != CURLE_OK)
+                return -EIO;
+
+        request_header = curl_slist_append(request_header, "Metadata:true");
+        if (!request_header)
+                return log_oom();
+
+        if (curl_easy_setopt(curl, CURLOPT_HTTPHEADER, request_header) != CURLE_OK)
+                return -EIO;
+
+        if (curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, network_cloud_meta_data_write_callback) != CURLE_OK)
+                return -EIO;
+
+        if (curl_easy_setopt(curl, CURLOPT_WRITEDATA, m) != CURLE_OK)
+                return -EIO;
+
+        if (curl_easy_perform(curl) != CURLE_OK)
```

Comment:
this can block indefinitely, so I think we need a timeout at the very least (curl_easy_setopt(curl, CURLOPT_TIMEOUT, ...))

Reply:
yes right